### ISSUE TITLE
Possibly a memory leak

### ISSUE THREAD

**Issue by AMD-NICK** [lunarmodules/copas#151]
For the last week I've been trying to find the causes of memory leaks in my application code from morning till night. After about 24 hours of running my bot in Lua, `collectgarbage("count")` reaches more than a gigabyte (lots of http requests running). I still don't really understand how garbage collection works in Lua, but maybe some coroutines aren't unloaded from memory or whatever. (Of course, I'm not excluding that the problem is in my code, but I've already searched everything)

Below is what the leak looks like on different versions of Lua. It always happens about the same way, in steps ( strange )

![](https://i.imgur.com/9uX2l5D.png)

I came to standalone Lua after Garry's Mod. It has a function [`timer.Simple(iDelay, fCallback)`](https://wiki.facepunch.com/gmod/timer.Simple) for "one-time" timers and [`timer.Create(sName, iDelay, iReps, fCallback)`](https://wiki.facepunch.com/gmod/timer.Create) for timers with repetition.

Below is code that in Garry's Mod runs very fast and leaves almost no garbage behind. More below is similar code for copas (`Lua 5.3.6`), which runs longer and leaves behind ~128 MB of garbage on Copas version 4.0.0 and ~96 MB on version 3.0.0

```lua
local function print_garbage()
	collectgarbage("collect")
	print("count", collectgarbage("count") / 1024)
end

timer.Create("garbage_timer", 0.5, 0, print_garbage)

print_garbage()

for i = 1,1000000 do
	timer.Simple(0.1, function() end)
end
```

`tests/timers_leak.lua`

```lua
package.path = string.format("../src/?.lua;%s", package.path)

local function print_garbage()
	collectgarbage("collect")
	print("count", collectgarbage("count") / 1024)
end

local co_timer = require("copas.timer")

co_timer.new({
	name      = "garbage_timer",
	delay     = 0.5,
	recurring = true,
	callback  = print_garbage,
})

print_garbage()

for i = 1,1000000 do
	-- timer.Simple(0.1, function() end)
	co_timer.new({
		name      = "name" .. i,
		delay     = 0.1,
		recurring = false,
		callback  = function(self)

		end,
	})
end

require("copas").loop()
```

I gave an example with timers, but I assume that it's not so much about them as it is about coroutines or that somewhere they are not unloaded from memory. Or maybe it's not a copas problem at all, but a specific feature of lua

## Steps to reproduce

```bash
git clone https://github.com/lunarmodules/copas.git copas_leak_test
cd copas_leak_test/tests
curl -o timer_leak.lua https://pastebin.com/raw/b81qvt3S
lua timer_leak.lua
```

**Comment by Tieske** [comment]
Interesting find. I was investigation CPU usage, which led to inexplicable timings in my applications. Some steps would take multiple seconds occasionally.

A memory leak might very well explain this, since I find that timings slowly rise to an extreme value, and then sharply drop again. Also the problem typically appears in the "core-timer" task, but not always.

some logs (only tasks that take more than 1 second before yielding are logged);
```
Sun Oct 30 21:06:07 2022 ERROR task ran for 1002 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 21:05:07 2022 ERROR task ran for 1003 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 21:04:06 2022 ERROR task ran for 1000 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 21:03:07 2022 ERROR task ran for 1003 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 21:02:07 2022 ERROR task ran for 1001 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 21:01:06 2022 ERROR task ran for 1001 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 21:00:07 2022 ERROR task ran for 1003 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 20:58:06 2022 ERROR task ran for 1000 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 20:56:07 2022 ERROR task ran for 1000 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 19:16:07 2022 ERROR task ran for 1009 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:47:40 2022 ERROR task ran for 5197 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:47:35 2022 ERROR task ran for 383403 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:41:11 2022 ERROR task ran for 16440 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:40:55 2022 ERROR task ran for 1146218 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:21:48 2022 ERROR task ran for 191422 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:18:37 2022 ERROR task ran for 86991 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:17:10 2022 ERROR task ran for 64103 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:15:57 2022 ERROR task ran for 51903 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:14:49 2022 ERROR task ran for 43820 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:13:43 2022 ERROR task ran for 37837 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:12:39 2022 ERROR task ran for 33297 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:11:35 2022 ERROR task ran for 29791 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:10:33 2022 ERROR task ran for 27063 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:09:30 2022 ERROR task ran for 24639 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:08:28 2022 ERROR task ran for 22713 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:07:27 2022 ERROR task ran for 21113 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:06:25 2022 ERROR task ran for 19620 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:05:24 2022 ERROR task ran for 18346 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:04:23 2022 ERROR task ran for 17283 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:03:22 2022 ERROR task ran for 16360 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:02:21 2022 ERROR task ran for 15480 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:01:20 2022 ERROR task ran for 14686 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 17:00:20 2022 ERROR task ran for 13977 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:59:19 2022 ERROR task ran for 13375 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:58:18 2022 ERROR task ran for 12812 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:57:18 2022 ERROR task ran for 12320 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:56:17 2022 ERROR task ran for 11783 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:55:17 2022 ERROR task ran for 11360 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:54:16 2022 ERROR task ran for 10934 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:53:16 2022 ERROR task ran for 10538 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:52:16 2022 ERROR task ran for 10200 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:51:15 2022 ERROR task ran for 9878 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:50:15 2022 ERROR task ran for 9572 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:49:15 2022 ERROR task ran for 9264 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:48:15 2022 ERROR task ran for 8968 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:47:14 2022 ERROR task ran for 8726 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:46:14 2022 ERROR task ran for 8497 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:45:14 2022 ERROR task ran for 8244 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:44:14 2022 ERROR task ran for 8048 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:43:13 2022 ERROR task ran for 7833 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:42:13 2022 ERROR task ran for 7665 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:41:13 2022 ERROR task ran for 7457 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:40:13 2022 ERROR task ran for 7281 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:39:13 2022 ERROR task ran for 7131 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:38:12 2022 ERROR task ran for 6991 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:37:12 2022 ERROR task ran for 6799 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:36:12 2022 ERROR task ran for 6680 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:35:12 2022 ERROR task ran for 6555 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:34:12 2022 ERROR task ran for 6424 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:33:12 2022 ERROR task ran for 6279 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:32:12 2022 ERROR task ran for 6158 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:31:12 2022 ERROR task ran for 6048 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:30:11 2022 ERROR task ran for 5949 milliseconds. (coroutine: copas_core_timer, socket: nil)
Sun Oct 30 16:29:11 2022 ERROR task ran for 5842 milliseconds. (coroutine: copas_core_timer, socket: nil)
```

**Comment by Tieske** [comment]
Did a quick test, but your test code is flawed. The size increase you're seeing is because of the control structures growing.

The way this works is that a Lua table has a certain "capacity" when first created, when more elements are added, beyond the capacity, another internal structure is allocated to hold these additional values. This repeats itself as the number of entries grow. Once the table is let go of, the GC will collect it. and memory will be freed. 
In this case, that doesn't happen. That's because the internal copas control structures (holding all the tasks and timers), will grow to hold 1 million timers, but the tables themselves are never GC'ed because Copas keeps using them over its lifetime.

So the memory is allocated only for very large, but empty, tables. There's little to be done about that. The way to check this, is to create 1 million timers, way for the GC cycles, and then repeat creating 1 million timers. If you do this repetitively, you'll see that memory usage stabilizes. (because the control structures no longer grow, since the limit of 1 million was reached the first time).

But I adapted your test code and I think I found the leak. It looks like it is timeout related.
```lua
package.path = string.format("../src/?.lua;%s", package.path)

local copas = require("copas")
local co_timer = require("copas.timer")

local function print_garbage()
  collectgarbage()
  print("memory use: ".. tostring(collectgarbage("count") / 1024) .. " Mb")
end


local test_done = false

co_timer.new({
  name      = "garbage_timer",
  delay     = 0.5,
  recurring = true,
  callback  = function(tmr)
    print_garbage()
    if test_done then
      tmr:cancel() -- cancel myself (this timer) to exit
    end
  end,
})


local function single_test()

  print("======= creating threads")

  for i = 1,500000 do
    copas.addnamedthread("task " .. i, function()
      copas.pause(0)
    end)
  end

  copas.pause(20)

  print("======= creating timeouts")

  local lock = require("copas.lock").new()
  for i = 1,500000 do
    copas.addnamedthread("timer " .. i, function()
      lock:get()
      copas.pause(0)
      lock:release()
    end)
  end

  copas.pause(20)
end

copas(function()
  local runs = 20
  for i = 1, runs do
    single_test()
    print(runs - i, "still to go")
  end
  test_done = true
end)
```
After every run, the memory stabilizes, and only after creating the timeouts, the level at which it stabilizes is going up. For every run.

**Comment by AMD-NICK** [comment]
So there is no way to fix this problem with Lua? Maybe by recreating some table?

**Comment by AMD-NICK** [comment]
I looked inside these variables while the script worked and there's a lot of data showing up. Maybe this could be the cause of the problem?

https://github.com/Tieske/timerwheel.lua/blob/master/src/timerwheel/init.lua#L87

https://github.com/Tieske/timerwheel.lua/blob/master/src/timerwheel/init.lua#L80

**Comment by Tieske** [comment]
can you try and see what happens if you comment out these 2 lines: 

https://github.com/Tieske/timerwheel.lua/blob/cfee1b8fa1cd8054bab0abd070270f8a5e939e48/src/timerwheel/init.lua#L137-L138

**Comment by Tieske** [comment]
created a fix, timerwheel version 1.0.2.

Please give that a try. Hopefully it fixes the issue.

**Comment by AMD-NICK** [comment]
Testing. I will report the results in about 24 hours

**Comment by AMD-NICK** [comment]
Preliminary chart.

Left luavela (ujit). 200 mbytes of RAM in 1:40 hours of work.

In the center is luajit from tarantool. Wasting 145 MB of RAM for the same amount of time.

On the right is the current situation. Memory waste is about 18 MB (on luavela, but with fix)

You can see that memory is also wasted, but the situation is much better. I will keep watching

<img width="1417" alt="image" src="https://user-images.githubusercontent.com/9200174/199885266-04f07308-ee89-442b-99a5-b017985d121d.png">

**Comment by AMD-NICK** [comment]
It's incredible result, but looks like this tables still collecting garbage. I have restarted the script and made timerwheel object with rings field global again to get more info. I will keep watching and collect information

P.S. Garbage graph for runtime period. Non-fixed timerwheel at the left: 
<img width="1406" alt="screenshot_2022-11-05_05 31 53@2x" src="https://user-images.githubusercontent.com/9200174/200099078-7aefbdff-d0bb-42e7-8e8d-2bfc8d5414c6.png">

**Comment by AMD-NICK** [comment]
I changed something before the restart and now I don't see the leak 🤔

**Comment by AMD-NICK** [comment]
<img width="1481" alt="image" src="https://user-images.githubusercontent.com/9200174/200155966-6e3451a3-da7f-4816-b676-58ba35a3cfcc.png">

Yes, the problem is definitely solved. I'm so glad. For about a month I searched for the cause and finally the problem is gone. Thank you for your help and patience ❤️

In case you're interested, I found the leak thanks to the memory profiler that's built into [Tarantool's luajit](https://github.com/tarantool/luajit). Without it I would have lost my mind. Here is the [documentation](https://www.tarantool.io/en/doc/latest/reference/tooling/luajit_memprof/)

**Comment by Tieske** [comment]
I finally tracked down the last culprit in the memory and CPU issues. Fixed in version 4.6.0 of Copas. 

Had to do with queue and semaphores using a 10yr timeout as a substitute for "never timeout". This caused the timerwheel to create a very large structure, and slowly filling it with tables. Eventually the memory would be released but the allocation over time would be way to high. In a cleanup cycle it would process the entire structure and create a "stop-the-world" event as such, which could take 45 to 60 seconds (in my test case).

The fix is to use `math.huge`, which will now cause Copas to NOT schedule a timeout.

Closing this.
(finally 😥 )
